### ISSUE TITLE
[proposal] please avoid to change protected/public method names

### ISSUE THREAD

**Issue by alberto-bottarini** [laravel/ideas#584]
Hi, I just found a bug on my code. This bug is caused by the change of a method name from Laravel 5.3 to Laravel 5.4 in ```Illuminate\Database\Eloquent\Relations\BelongsToMany``` class.

L53 has a ```setJoin``` method. L54 renames it to ```performJoin```.

In my project i had to extends the standard behaviour of this relation and I override setJoin method and upgrading to L54 this methods stops to being overridden because of the name changing. 
Also this changes is not tracked in any L54 changelog so I didn't have a way to prevent this problem.

What do you think?

**Comment by deleugpn** [comment]
Breaking changes will always happen during minor releases. Are you proposing the framework restricts itself from evolving? The fact that you can easily fix this in 2 minutes makes it a weak argument against breaking changes.

**Comment by alberto-bottarini** [comment]
Should we consider method renaming as a "evolving" process? It gives a very small advantage but could create annoying problems.

**Comment by alberto-bottarini** [comment]
I could agree with you about renaming public methods in order to improve eloquency. But these changes should be reported.

**Comment by sisve** [comment]
That one line of code was probably fixed in 2 minutes ... if he knew the new name. It's not in the changelog, so he didn't know it. So I'll presume walking through git commits and viewing histories to find the code line. I knew what I was looking for, so I found the commit https://github.com/laravel/framework/commit/7f4dd21a7543009238e7cc1fe3820ab970397e52 in a few minutes... but I am still required to understand the changes in the code to understand that setJoin is now performJoin. I do not consider it clear from that code. But okay, I'll asume that 15 minutes will be spent on the issue...

Now, this is one user. Imagine that one promille of the users (I made that number up) runs into this. That becomes 175 sites if I am to believe https://trends.builtwith.com/framework/Laravel. This puts us at 175*15 minutes = about 44 hours.

Now, this is only one of ten (another number I made up) renamed methods not tracked by the changelog. We're now up to 18 days of developer time spent on these small "you can easily fix this in 2 minutes".

And then there are the Stack Overflow posts that are broken, the tutorials that are broken, the Laracasts answers that are broken, ...

And this issue, and future issues, and breakage and misery and pain would be non-existent if old methods were left in place with a @deprecated annotation telling the new method name for one release.  Yes, this require a decent IDE that shows these notices.

Or perhaps keep the old methods to throw an exception pointing out the new methods?

Relevant: https://github.com/laravel/internals/issues/383 (Deprecating changes for 1 release instead of removing)

**Comment by deleugpn** [comment]
The argument that this is not documented is great and perhaps a good approach would be to quickly PR the docs to add this so future developers don't end up spending 44 hours on it. I'm just completely close-minded regarding this issue's title. I read it as "Don't change the framework because I'm using it, thanks.".

**Comment by alberto-bottarini** [comment]
@sisve you completely get the point. I had a bug in production for some days because my overwritten setJoin was ignored.

@deleugpn title was intentionally provocative and rude. I don't think that was a mistake, it's not the first time that i get this kind of error upgrading Laravel. Basically method changes don't get reported.

**Comment by joshmanders** [comment]
It's kind of important to list breaking changes, in the changelog, no?

The whole "It's only 2 minute fix" is not only a bunk excuse but also a reckless attitude to have.

**Comment by alberto-bottarini** [comment]
Tecnically speaking, a protected method name change is a breaking change?

**Comment by scrwdnet** [comment]
I agree with @alberto-bottarini in that it probably should have been recorded in the changelog but this just screams out about a lack of testing of the individual project. You are bumping a version and just did `composer update && git push master` <-- @joshmanders **that** is reckless.

I'll just leave this here: https://adamwathan.me/test-driven-laravel/

Don't rely on other people to be 100% perfect every day, test your code especially before pushing to production.

Has anyone PR'd the docs yet to document this change - it is open source after all 😄

**Comment by sisve** [comment]
@alberto-bottarini Yes, it is. Everyone can derive from the method and override it. It's part of the _public interface of the class_, it's just not a public method.

A good way to figure out if something is part of the public interface is to check if you can access the property/method without reflection. This includes subclasses. If you can subclass something and override a method, then someone probably has and it is a breaking change to rename it. This is also why protected methods needs tests; they are part of the public api.

If you do not want people messing with your class members (and be free to rename them at any time) make them private.

I like to link to Symfony's [Backward Compatibility Promise](https://symfony.com/doc/current/contributing/code/bc.html) when writing about this.

**Comment by joshmanders** [comment]
@scrwdnet few years ago a group of people you may have heard of, they founded the site we're commenting on right now.. So they came up with this idea called Semantic Versioning, a way for computers to know when to update things, and what will break and all that... You see this program we use to manage our dependencies also utilizes it... So if I do `^5.4.*` that tells composer that I want all updates that are `5.4` or higher, but less than `6.0` because according to semver, I should not get any breaking changes, only new features and bug fixes.

Now you're about to say that Laravel doesn't follow semver, and that's fine... But if you're not gonna, you should be pretty diligent on making sure that your users know when stuff is gonna break, and what broke...

Yes I agree that `composer update && git push master` is reckless, but putting the blame on your users for not knowing what you changed is a shitty way to remedy it.

**Comment by scrwdnet** [comment]
@joshmanders 100% agree but this is **free** open source software and shit happens. Nobody blamed the users (well maybe @deleugpn, but I think that was reactive to the, admittedly, 'provocative & rude' title).

I feel my point is still valid that you (not *you*...) have made a choice to bump a version and don't have coverage and an error has made it into production (I can only imagine not the first time if you don't have the coverage). The first thing you did to an open source project was to open an issue with a douche title and not even open a PR to help others out who might have the same issue.

Here are some optional titles you (still not *you*...) could have used (along side a PR to the docs):
- [proposal] improve recording changes to public/protected method names
- [proposal] discuss possibility of implementing a backwards compatibility promise like Symphony

We need more PDD™ (positivity driven development) or NADDD™ (not a douche driven development). Laravel is a great tool but like I said previously people aren't always 100% correct in everything you have to take responsibility for your own actions at some point and stop playing the blame game. You want a change then muck in.

That's my two cents/thoughts/opinions anyway...

**Comment by LKDevelopment** [comment]
@joshmanders you know that laravel doesn't follow semver? https://laravel.com/docs/5.4/releases#versioning-scheme

**Comment by deleugpn** [comment]
@LKDevelopment I think he does since he said 

> Now you're about to say that Laravel doesn't follow semver, and that's fine...

**Comment by LKDevelopment** [comment]
@deleugpn Sure, but that it doesn't follow semver is readable on the documentation since a long time :) Was just a note that is listed in the documentation.

**Comment by joshmanders** [comment]
@LKDevelopment 
> But if you're not gonna, you should be pretty diligent on making sure that your users know when stuff is gonna break, and what broke...

**Comment by LKDevelopment** [comment]
@joshmanders sure, the changelog for every laravel version isn't so good. It shouldn't be difficult to write something that gets the commits for each version. I'll look to do something like this to help the Community

**Comment by scrwdnet** [comment]
*From the docs:*
> Why Doesn't Laravel Use Semantic Versioning?
>
>On one hand, all optional components of Laravel (Cashier, Dusk, Valet, Socialite, etc.) do use semantic versioning. However, the Laravel framework itself does not. The reason for this is because semantic versioning is a "reductionist" way of determining if two pieces of code are compatible. Even when using semantic versioning, you still must install the upgraded package and run your automated test suite to know if anything is actually incompatible with your code base.
>
>So, instead, the Laravel framework uses a versioning scheme that is more communicative of the actual scope of the release. Furthermore, since patch releases never contain intentional breaking changes, you should never receive a breaking change as long as your version constraints follow the  paradigm.minor.* convention.

*TL;DR*
Test your applications...

**Comment by joshmanders** [comment]
TL;DR

If you submit a PR adding a breaking change, also add a commit to the changelog telling about that breaking change.

Look at that, so now when people upgrade and run their tests and they fail, they can be like "this worked before, let me check the changelog... Oh look there it is, quick change woohooo"

**Comment by scrwdnet** [comment]
@LKDevelopment I think Laravel Shift does something with a big diff accumulator? Could be wrong though.

**Comment by scrwdnet** [comment]
@joshmanders agree as before but again there is a reliance on other people.

Gotta take that responsibility at some point.

**Comment by alberto-bottarini** [comment]
Thank @deleugpn for the PR. 

I did something interesting (and borind). I made a ```git diff``` between 5.4 and 5.3 on ```BelongsToMany.php``` and I found:

- protected property ```otherKey``` renamed to ```relatedKey```
- protected method ```setJoin``` renamed to ```performJoin```
- protected method ```setWhere``` renamed to ```addWhereContraints```
- protected method ```getOtherKey``` renamed to ```getQualifiedRelatedKeyName```
- protected method ```getForeignKey``` renamed to ```getQualifiedForeignKeyName```

and some other differences too tricky to understand. 

These in only *one* file/class. I'm not sure if reporting these changes in changelog could be useful.
